### PR TITLE
Add parameter validation for the DHM module

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ Based on #1643, this PR implements parameter validation for the DHM module.

### CODE REVIEW

**Comment by hanno-arm** [comment]
@mpg As mentioned in the bignum PR, I don't think we should make parameter validation too fine-grained, as it is unavoidably only an approximation to the preconditions. I'd therefore prefer to stick to the primitive non-NULL check. What do you think?

File: library/dhm.c [hunk 1]
```diff
@@ -103,6 +108,7 @@ static int dhm_check_range( const mbedtls_mpi *param, const mbedtls_mpi *P )
     mbedtls_mpi L, U;
     int ret = 0;
 
+
```
Comment (AndrzejKurek) [inline library/dhm.c]: Unnecessary line

File: tests/suites/test_suite_dhm.function [hunk 2]
```diff
@@ -7,6 +7,112 @@
...
+                                                     &len, rnd_std_rand,
+                                                     NULL ) );
+    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_DHM_BAD_INPUT_DATA,
+                            mbedtls_dhm_calc_secret( &ctx, buf, 42,
+                                                     NULL, rnd_std_rand,
+                                                     NULL ) );
+
+#if defined(MBEDTLS_ASN1_PARSE_C)
+    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_DHM_BAD_INPUT_DATA,
+                            mbedtls_dhm_parse_dhm( NULL, buf, 42 ) );
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_dhm.function]: Please refactor the "42" constant.

Comment (hanno-arm) [reply tests/suites/test_suite_dhm.function]: Done.

File: library/dhm.c (hunk 1 above)
Comment (hanno-arm) [reply library/dhm.c]: Fixed.

**Comment by hanno-arm** [comment]
@AndrzejKurek @mpg Please re-review.

**Comment by sbutcher-arm** [comment]
@hanno-arm - Can you please rebase? Thanks

**Comment by hanno-arm** [comment]
Rebased on top of #1643.

File: include/mbedtls/dhm.h [hunk 3]
```diff
@@ -143,31 +149,37 @@ void mbedtls_dhm_init( mbedtls_dhm_context *ctx );
...
+ *                 and have the DHM parameters set. It may or may not
+ *                 already have imported the peer's public key.
  * \param x_size   The private key size in Bytes.
- * \param olen     The number of characters written.
- * \param output   The destination buffer.
- * \param f_rng    The RNG function.
- * \param p_rng    The RNG context.
+ * \param olen     The address at which to store the number of Bytes
+ *                 written on success. This must not be \c NULL.
+ * \param output   The destination buffer. This must be a writable buffer of
```
Comment (gilles-peskine-arm) [inline include/mbedtls/dhm.h]: Not a blocker for this PR, but when I read this sentence, I don't get a clear idea of how to calculate the buffer size. I don't expect users of the library to understand what “the reduced binary presentation” (shouldn't this be “representation”?) is.

File: include/mbedtls/dhm.h [hunk 4]
```diff
@@ -143,31 +149,37 @@ void mbedtls_dhm_init( mbedtls_dhm_context *ctx );
...
- *
- * \note           The destination buffer must be large enough to hold
- *                 the reduced binary presentation of the modulus, the generator
- *                 and the public key, each wrapped with a 2-byte length field.
- *                 It is the responsibility of the caller to ensure that enough
- *                 space is available. Refer to \c mbedtls_mpi_size to computing
- *                 the byte-size of an MPI.
+ * \brief          This function generates a DHM key pair and exports its
+ *                 public part together with the DHM parameters in the format
+ *                 used in a TLS ServerKeyExchange handshake message.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/dhm.h]: Non-blocker: The functions that refer to a TLS ServerKeyExchange message should refer to the RFC section that defines it.

**Comment by sbutcher-arm** [comment]
PR has been reviewed, approved, passes CI, but needs rebasing! 

PR #1643 has now been merged, so removing the 'needs: preceding PR' label.

**Comment by mpg** [comment]
@sbutcher-arm I don't think this branch needs rebasing, it has no conflicts and was based on a recent enough version to pass CI.

**Comment by sbutcher-arm** [comment]
@mpg - Agreed! I just saw duplicate commits from #1643, and assumed it needed it.

Jenkins CI passes in all but the BSD timing test, reviewed and approved, so labelling as 'ready to merge'.

**Pull request merged**